pkg/ingester/ingester.go

```diff
@@ -502,6 +545,55 @@ func (i *Ingester) append(ctx context.Context, userID string, labels labelPairs,
 	return err
 }
 
+func (i *Ingester) appendMetadata(userID string, m *client.MetricMetadata) error {
+	i.userStatesMtx.RLock()
+	if i.stopped {
+		i.userStatesMtx.RUnlock()
+		return fmt.Errorf("ingester stopping")
+	}
+	i.userStatesMtx.RUnlock()
+
+	i.usersMetadataMtx.Lock()
+	userMetadata := i.getOrCreateUserMetadata(userID)
+	i.usersMetadataMtx.Unlock()
+
+	if err := userMetadata.add(m.GetMetricName(), m); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (i *Ingester) getOrCreateUserMetadata(userID string) *userMetricsMetadata {
+	userMetadata, ok := i.usersMetadata[userID]
+	if !ok {
+		userMetadata = newMetadataMap(i.limiter, i.metrics, userID)
+		i.usersMetadata[userID] = userMetadata
+	}
+
+	return userMetadata
+}
+
+func (i *Ingester) purgeUserMetricsMetadata() {
```

Comment:
We shouldn't take the global lock for the entire purge duration. It can be re-implemented like this:

```

func (i *Ingester) getOrCreateUserMetadata(userID string) *userMetricsMetadata {
	userMetadata := i.getUserMetadata(userID)
	if userMetadata != nil {
		return userMetadata
	}

	i.usersMetadataMtx.Lock()
	defer i.usersMetadataMtx.Unlock()

	// Ensure it has not been created in the meanwhile.
	userMetadata, ok := i.usersMetadata[userID]
	if !ok {
		userMetadata = newMetadataMap(i.limiter, i.metrics, userID)
		i.usersMetadata[userID] = userMetadata
	}

	return userMetadata
}

func (i *Ingester) getUserMetadata(userID string) *userMetricsMetadata {
	i.usersMetadataMtx.RLock()
	defer i.usersMetadataMtx.RUnlock()

	return i.usersMetadata[userID]
}

func (i *Ingester) getUsersWithMetadata() []string {
	i.usersMetadataMtx.RLock()
	defer i.usersMetadataMtx.RUnlock()

	userIDs := make([]string, 0, len(i.usersMetadata))
	for userID, _ := range i.usersMetadata {
		userIDs = append(userIDs, userID)
	}

	return userIDs
}

func (i *Ingester) purgeUserMetricsMetadata() {
	deadline := time.Now().Add(-i.cfg.MetadataRetainPeriod)

	for _, userID := range i.getUsersWithMetadata() {
		metadata := i.getUserMetadata(userID)
		if metadata == nil {
			continue
		}

		// Remove all metadata that we no longer need to retain.
		deletedCount := metadata.purge(deadline)

		i.metrics.memMetadata.Sub(float64(deletedCount))
		i.metrics.memMetadataRemovedTotal.WithLabelValues(userID).Add(float64(deletedCount))
	}
}
```

**SUPER IMPORTANT**: with the lock optimizations I'm suggesting, there's no safe way to do `delete(i.usersMetadata, userID)` once the user metadata is empty (because another one outside the global lock could retain a reference to it and writing). It's perfectly OK keeping the data structure. It's super small and will not go over the max number of tenants in the cluster.